### PR TITLE
Reinstate #93800

### PR DESCRIPTION
**Pull request by b-naber**
https://github.com/rust-lang/rust/pull/93800 caused a regression in an alt builder with parallel enabled. https://github.com/rust-lang/rust/pull/94205 reverted that PR because of the regression. For an unknown reason the regression has disappeared, so we reinstate the changes in https://github.com/rust-lang/rust/pull/93800 here.

r? @Mark-Simulacrum

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -5040,9 +5040,9 @@ dependencies = [
 
 [[package]]
 name = "textwrap"
-version = "0.14.2"
+version = "0.15.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "0066c8d12af8b5acd21e00547c3797fde4e8677254a7ee429176ccebbe93dd80"
+checksum = "b1141d4d61095b28419e22cb0bbf02755f5e54e0526f97f1e3d1d160e60885fb"
```
Comment (Mark-Simulacrum) [inline Cargo.lock:5045]: These changes look unrelated?

Comment (b-naber) [reply Cargo.lock:5045]: Not sure, thought I had to include the changes to `Cargo.lock`. I've had problems in the past by not including changes to that file. What am I supposed to do here?

Comment (Mark-Simulacrum) [reply Cargo.lock:5045]: Since this PR doesn't appear to update any Cargo.toml files, you should be able to drop these changes by resetting from master; e.g. `git restore --source=origin/master -- Cargo.lock`.

**Comment by Mark-Simulacrum** [comment]
The original PR also had some mir-opt test updates, but it looks like they're not included here. Is this change different? Did something change on master?

**Comment by b-naber** [comment]
> The original PR also had some mir-opt test updates, but it looks like they're not included here. Is this change different? Did something change on master?

I was surprised by that as well. I did bless the tests in `mir-opt` though, but there were no changes detected. I'm not sure whether anything else changed that caused the changes from the previous PR to be omitted.

@lcnr and @oli-obk Do you have any idea why there aren't any diffs for the `mir-opt` tests?

**Comment by lcnr** [comment]
we changed mir dumps in #94209, so it's expected that there are no changes to that anymore here

**Comment by Mark-Simulacrum** [comment]
In that case I think we're likely prepared to move ahead here -- it would be nice to add a little more detail to the PR description reflecting the reasoning for the revert and the seemingly fixed regression (though through no actual changes).

r? @lcnr or @oli-obk on the actual code, though I presume that it's the same and likely still fine.

File: compiler/rustc_middle/src/thir/visit.rs [hunk 2]
```diff
@@ -123,7 +123,7 @@ pub fn walk_expr<'a, 'tcx: 'a, V: Visitor<'a, 'tcx>>(visitor: &mut V, expr: &Exp
         }
         Closure { closure_id: _, substs: _, upvars: _, movability: _, fake_reads: _ } => {}
         Literal { literal, user_ty: _, const_id: _ } => visitor.visit_const(literal),
-        StaticRef { literal, def_id: _ } => visitor.visit_const(literal),
+        StaticRef { .. } => {}
```
Comment (lcnr) [inline compiler/rustc_middle/src/thir/visit.rs:126]: please explicitly mention the ignored fields here

**Comment by lcnr** [comment]
@bors r+ rollup=never

i guess we want to be safe here
